### PR TITLE
Add close method to `openmc.StatePoint` and update docs/examples for best practices

### PR DESCRIPTION
**Pull request by pshriwise**
## Movitation

A recurring issue in the forum is users running into the error

```shell
RuntimeError: Failed to open HDF5 file with mode 'w': summary.h5
```

when working with OpenMC's Python API. This is usually related to an open StatePoint file when OpenMC is being executed. The open HDF5 file handle prevents an overwrite of the `summary.h5` file (auto-linked to the `StatePoint` object by default) and causes OpenMC to crash. This can also happen if OpenMC tries to overwrite a statepoint file with an open handle, so we can see this even if writing the summary file is disabled.

A couple of forum discussions related to this issue:

  - https://openmc.discourse.group/t/plot-tally-result-when-using-distribcellfilter/1247/4
  - https://openmc.discourse.group/t/error-in-opening-h5-file/288

## Changes

This PR adds a `openmc.StatePoint.close` method that closes the HDF5 files related to the statepoint object. `StatePoint` file handles are most easily managed using context managers when extracting data, but there are times when these regions may become unreasonably large. The `StatePoint.close` method is meant to handle these cases.

To increase user awareness, I've also updated the examples to use context managers when possible and always close the `StatePoint` objects as a matter of best practice if needed.

### CODE REVIEW

File: docs/source/usersguide/troubleshoot.rst [hunk 1]
```diff
@@ -45,6 +45,23 @@ with the :envvar:`OPENMC_CROSS_SECTIONS` environment variable. It is recommended
 to add a line in your ``.profile`` or ``.bash_profile`` setting the
 :envvar:`OPENMC_CROSS_SECTIONS` environment variable.
 
+RuntimeError: Failed to open HDF5 file with mode 'w': summary.h5
+****************************************************************
+
+This often occurs when working with the Python API and executing multiple OpenMC
+runs in a script. If an :class:`openmc.StatePoint` is open in the Python interpreter,
+the file handle of the statpoint file as well as the linked `summary.h5` file will
```
Comment (paulromano) [inline docs/source/usersguide/troubleshoot.rst:53]:
```suggestion
the file handle of the statepoint file as well as the linked `summary.h5` file will
```

File: docs/source/usersguide/troubleshoot.rst [hunk 2]
```diff
@@ -45,6 +45,23 @@ with the :envvar:`OPENMC_CROSS_SECTIONS` environment variable. It is recommended
...
+be unavailable for writing, causing this error to appear. To avoid the situation,
```
Comment (paulromano) [inline docs/source/usersguide/troubleshoot.rst:54]:
```suggestion
be unavailable for writing, causing this error to appear. To avoid this situation,
```

File: examples/jupyter/candu.ipynb [hunk 3]
```diff
@@ -399,688 +399,9 @@
...
-     "metadata": {},
-     "output_type": "execute_result"
-    }
-   ],
-   "source": [
-    "output_tally = sp.get_tally()\n",
-    "output_tally.get_pandas_dataframe()"
+    "with openmc.StatePoint('statepoint.{}.h5'.format(settings.batches)) as sp:\n",
+    "    output_tally = sp.get_tally()\n",
+    "    output_tally.get_pandas_dataframe()"
```
Comment (paulromano) [inline examples/jupyter/candu.ipynb:404]: This no longer shows the dataframe. I think you'll need something like:
```
with blahblah:
    output_tally = sp.get_tally()
    df = output_tally.get_pandas_dataframe()
df
```

File: examples/jupyter/mdgxs-part-i.ipynb [hunk 4]
```diff
@@ -1141,6 +1137,31 @@
...
+     },
+     "execution_count": 22,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "dr_tally.filters"
+   ]
+  },
```
Comment (paulromano) [inline examples/jupyter/mdgxs-part-i.ipynb:1164]: Was this addition intentional?

File: examples/jupyter/mg-mode-part-i.ipynb [hunk 5]
```diff
@@ -205,7 +197,16 @@
...
    "execution_count": 7,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/home/shriwise/opt/openmc/openmc/openmc/surface.py:1511: FutureWarning: \"ZCylinder(...) accepts an argument named 'r', not 'R'. Future versions of OpenMC will not accept the capitalized version.\n",
+      "  FutureWarning)\n"
```
Comment (paulromano) [inline examples/jupyter/mg-mode-part-i.ipynb:206]: While you're at it, do you mind addressing this?

**Review by paulromano — CHANGES_REQUESTED** [review]
Good call on adding this method and updating the examples. Also, thanks for adding something to our long-neglected troubleshooting guide :smile:

File: examples/jupyter/candu.ipynb (hunk 3 above)
Comment (pshriwise) [reply examples/jupyter/candu.ipynb:404]: Good catch! Thanks!

File: examples/jupyter/mdgxs-part-i.ipynb (hunk 4 above)
Comment (pshriwise) [reply examples/jupyter/mdgxs-part-i.ipynb:1164]: Nope. Not sure how that snuck in there. Thanks!

File: examples/jupyter/mg-mode-part-i.ipynb (hunk 5 above)
Comment (pshriwise) [reply examples/jupyter/mg-mode-part-i.ipynb]: Can do!

**Review by paulromano — APPROVED** [review]

**Comment by paulromano** [comment]
Thanks again @pshriwise!
